extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/devmode/DevConsoleCORSFilter.java

```diff
@@ -0,0 +1,51 @@
+package io.quarkus.vertx.http.runtime.devmode;
+
+import java.util.List;
+import java.util.Optional;
+
+import org.eclipse.microprofile.config.ConfigProvider;
+import org.jboss.logging.Logger;
+
+import io.quarkus.vertx.http.runtime.cors.CORSConfig;
+import io.quarkus.vertx.http.runtime.cors.CORSFilter;
+import io.vertx.core.Handler;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.core.http.HttpServerRequest;
+import io.vertx.core.http.HttpServerResponse;
+import io.vertx.ext.web.RoutingContext;
+
+public class DevConsoleCORSFilter implements Handler<RoutingContext> {
+    private static final Logger LOG = Logger.getLogger(DevConsoleCORSFilter.class);
+
+    private static final String HTTP_PORT_CONFIG_PROP = "quarkus.http.port";
+    private static final String LOCAL_HOST = "localhost";
+    private static final String HTTP_LOCAL_HOST = "http://" + LOCAL_HOST;
+    private static final String HTTPS_LOCAL_HOST = "https://" + LOCAL_HOST;
+
+    public DevConsoleCORSFilter() {
+    }
+
+    private static CORSFilter corsFilter() {
+        int httpPort = ConfigProvider.getConfig().getValue(HTTP_PORT_CONFIG_PROP, int.class);
+        CORSConfig config = new CORSConfig();
+        config.origins = Optional.of(List.of(HTTP_LOCAL_HOST + ":" + httpPort, HTTPS_LOCAL_HOST + ":" + httpPort));
+        return new CORSFilter(config);
+    }
+
+    @Override
+    public void handle(RoutingContext event) {
+        HttpServerRequest request = event.request();
+        HttpServerResponse response = event.response();
+        String origin = request.getHeader(HttpHeaders.ORIGIN);
+        if (origin == null) {
+            corsFilter().handle(event);
+        } else {
+            if (origin.startsWith(HTTP_LOCAL_HOST) || origin.startsWith(HTTPS_LOCAL_HOST)) {
+                corsFilter().handle(event);
+            } else {
```

Comment:
This is a **serious** security bug!

You're doing CORS validation wrong. If you're building on top of vertx-web, you should use the provider handler for this. Here's why the code is wrong:

The origin header can contain the following values:

* `null`
* a origin e.g.: `http://localhost`
* `"null"` the "null" string

If you're arriving from a different protocol (e.g.: file://...) the origin is the "null" string.

Now the origin, when the value is like `http://localhost`. In the code you're testing as a prefix, which means an attacker can run another server as: `http://localhost.mydomain.com`. In this case I (as an attacker) can escape the CORS sandbox here and introduce a CVE.

Given that you're dealing with `private network` origins, you should see that in order to protect it with evergreen browsers, you need to deal with `Private Network Access Preflight`: https://developer.chrome.com/blog/private-network-access-preflight/

Finally, note that parsing the `Origin` isn't this trivial, looking at the code, it expects `Origin: http://localhost`. However, a user agent may return `Origin: http://localhost:80` and this is still the same origin. An example for this is when a user explicitly types the port on the url bar.

Reply:
@pmlopes, Hi

- If `Origin` is null we delegate to Quarkus CORS filter to deal with it.
- If Origin is "null" - we reject it

>Now the origin, when the value is like http://localhost. In the code you're testing as a prefix, which means an attacker can run another server as: http://localhost.mydomain.com. In this case I (as an attacker) can escape the CORS sandbox here and introduce a CVE.

No, it is not going to happen - this DevConsole is only double-protects and rejects any Origin which contains no `locahost` - but if someone posts `Origin: http://localhost.mydomain.com` then it will be rejected anyway - as we delegate to Quarkus CORS filter which is initialized with 2 concrete and only supported origins: `http://localhost:<quarkus.http.port>`, `https://localhost:<quarkus.http.port>`.

I'll update this PR to test that `http://localhost.mydomain.com` is rejected

>Finally, note that parsing the Origin isn't this trivial, looking at the code, it expects Origin: http://localhost. However, a user agent may return Origin: http://localhost:80 and this is still the same origin. An example for this is when a user explicitly types the port on the url bar.

See the prev comment.

Note - the goal of this CORS handler is not to provide a proper CORS support for DevConsole - the only goal is to reject Origins not containing either `http://localhost:<quarkus.http.port>` or `https://localhost:<quarkus.http.port>` - we do not want DevConsole endpoints be accessible from anywhere but the localhost

Reply:
I think we will need, in a second step, to replace the main CORS filter with the Vert.x one.

Reply:
@cescoffier Sure it can be reviewed